### PR TITLE
feature: add support for argoproj Rollouts 

### LINKED ISSUE #12: add support for argo rollouts
The controller-manager isn't compatible with argo Rollouts and needs a schema added? The Rollout 'wraps' the Deployment through a CRD so that it can control and manage multiple replicasets simultaneously for canary/b-g testing.

> 2022-03-25T14:17:13.572Z ERROR controller Reconciler error {"reconcilerGroup": "migrations.coderanger.net", "reconcilerKind": "Migrator", "controller": "migrator", "name": "admin", "namespace": "vela", "error": "error in user component reconcile: error finding template pod spec: error finding object type for owner reference &OwnerReference{Kind:Rollout,Name:admin-api,UID:6863d34c-8b0c-42b0-91a0-2d0f4da77b2f,APIVersion:argoproj.io/v1alpha1,Controller:*true,BlockOwnerDeletion:*true,}: no kind \"Rollout\" is registered for version \"argoproj.io/v1alpha1\" in scheme \"pkg/runtime/scheme.go:101\"", "errorVerbose": "no kind \"Rollout\" is registered for version \"argoproj.io/v1alpha1\" in scheme \"pkg/runtime/scheme.go:101\"\nerror finding object type for owner reference

### PR DESCRIPTION
**Pull request by zswanson**
Addresses #12 

- Adds a stub type for argoproj.io Rollout so that the migrator can understand how to find pod specs in a Rollout. In a discussion with @coderanger  it was mentioned we could avoid adding the deepcopy generation but the compiler insisted that it was a requirement. 🤷  
- Updated the controller to register the Rollout scheme and include that object type in its RBAC and decision making.
- Added a test (copy of the Deployment workload test) to ensure it correctly finds the pod templates in a Rollout.

~Did **not** add handling for the case where the Rollout is using`spec.workloadRef` as a reference to an existing Deployment.~ Added support for handling WorkloadRef (for Deployment only) after we had some other issues with managing kustomize strategic merges in Rollouts. This allows for cases where the Rollout spec only contains the Argo specific strategy and the pod template is in a referenced Deployment.

Co-authored with @tlarocque

### CODE REVIEW

**Comment by coderanger** [comment]
You should update the copyright info unless you're intentionally assigning that over to me :) We definitely need the deepcopy gen stuff but I think we can remove the controller-gen comments so it doesn't try to make a CRD? It might not anyway since this is in a different folder than normal.

**Comment by zswanson** [comment]
It didn't try to create a CRD that I saw, although at one point the makefile or something else managed to pull the remote argoproj rollout CRD spec into the source tree ... still not sure how that happened exactly. The deepcopy annotations were a first-time for us and were very confusing initially, we were just mimicking the migrator spec and then playing with it until it seemed to satisfy the compiler.

File: api/stubs/argoproj/groupversion_info.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//+kubebuilder:object:generate=true
+//+groupName=argoproj.io
```
Comment (zswanson) [inline api/stubs/argoproj/groupversion_info.go:16]: @coderanger are these generator comments you're thinking we can do without?

**Comment by zswanson** [comment]
Updated with support for the Rollout WorkloadRef

**Comment by zswanson** [comment]
@coderanger any other concerns with this? @jonasoneves and @tlarocque hav been running this branch for a few months now with ArgoRollouts

**Comment by coderanger** [comment]
I did the rearranging and merged it manually :)

The underlying issue is that as written, this was generating a CRD YAML for the stub which could easily lead to big mess. I've reworked the code layout to prevent that.

**Comment by coderanger** [comment]
Thanks for the contribution!

**Pull request merged**